File: go/bob/bob.go
package bob // package name must match the package name in bob_test.go

import (
	"strings"
	"unicode"
)

const testVersion = 2 // same as targetTestVersion

func Hey(s string) string {
	s = strings.TrimSpace(s)
	if s == "" {
		return "Fine. Be that way!"
	} else if strings.IndexFunc(s, func(r rune) bool { return unicode.IsLetter(r) }) != -1 &&
		strings.IndexFunc(s, func(r rune) bool { return unicode.IsLower(r) }) == -1 {
		return "Whoa, chill out!"
	} else if strings.HasSuffix(s, "?") {
		return "Sure."
	} else {
		return "Whatever."
	}
}

File: scala/bob/src/main/scala/bob.scala
class Bob {
  private def blank(s: String) = s.isEmpty

  private def shout(s: String) = {
    val s1 = s.toUpperCase
    val s2 = s.toLowerCase
    s == s1 && s1 != s2
  }

  private def question(s: String) = s.endsWith("?")

  def hey(s: String) = {
    val ts = s.trim
    if(blank(ts)) "Fine. Be that way!"
    else if(shout(ts)) "Whoa, chill out!"
    else if(question(ts)) "Sure."
    else "Whatever."
  }
}
